File: legate/pandas/core/bitmask.py
# Copyright 2021 NVIDIA Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from legate.pandas.common import types as ty
from legate.pandas.config import OpCode
from legate.pandas.core.pattern import Map, Projection


class Bitmask(object):
    alloc_type = ty.uint8

    def __init__(self, runtime, storage):
        self._runtime = runtime
        self._storage = storage
        self._compact_bitmask = None

    @property
    def region(self):
        return self._storage.region

    @property
    def field_id(self):
        return self._storage.field_id

    @property
    def dtype(self):
        return self._storage.dtype

    @property
    def num_pieces(self):
        return self._storage.num_pieces

    @property
    def launch_domain(self):
        return self._storage.launch_domain

    @property
    def storage(self):
        return self._storage.storage

    @property
    def ispace(self):
        return self._storage.ispace

    @property
    def primary_ipart(self):
        return self._storage.primary_ipart

    @property
    def fspace(self):
        return self._storage.fspace

    @property
    def compact_bitmask(self):
        if self._compact_bitmask is None:
            result = self._storage.storage.create_new_field(self.dtype)
            result.set_primary_ipart(self.primary_ipart)

            plan = Map(self._runtime, OpCode.TO_BITMASK)
            proj = Projection(self._storage.primary_ipart)

            plan.add_output_only(result)
            plan.add_scalar_arg(False, ty.bool)
            plan.add_scalar_arg(0, ty.uint32)

            plan.add_input(self._storage, proj)
            plan.add_scalar_arg(False, ty.bool)
            plan.add_scalar_arg(0, ty.uint32)

            plan.execute(self.launch_domain)

            self._compact_bitmask = result
        return self._compact_bitmask

    def clone(self):
        return Bitmask(self._runtime, self._storage.clone())

    def set_primary_ipart(self, ipart):
        self._storage.set_primary_ipart(ipart)

    def get_view(self, ipart):
        return self._runtime.create_region_partition(self.region, ipart)

    # XXX: This method should be used with caution as it gets blocked
    #      on a future internally
    def has_nulls(self):
        return self.count_nulls().sum().get_value() > 0

    def count_nulls(self):
        plan_count = Map(self._runtime, OpCode.COUNT_NULLS)
        proj = Projection(self._storage.primary_ipart)
        plan_count.add_input(self._storage, proj)
        return plan_count.execute(self.launch_domain).cast(ty.uint64)

    def to_raw_address(self):
        return self._storage.to_raw_address()

File: src/bitmask/bitmask.cc
/* Copyright 2021 NVIDIA Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include <assert.h>
#include <string.h>

#include "bitmask/bitmask.h"

namespace legate {
namespace pandas {

Bitmask::Bitmask(AllocType *bitmask, size_t num_elements)
  : bitmask(bitmask), num_elements(num_elements)
{
}

Bitmask::Bitmask(const AllocType *bitmask, size_t num_elements)
  : bitmask(const_cast<AllocType *>(bitmask)), num_elements(num_elements)
{
}

Bitmask::Bitmask(size_t num_elements, alloc::Allocator &allocator, bool init, bool init_value)
  : bitmask(allocator.allocate_elements<AllocType>(num_elements)), num_elements(num_elements)
{
  if (init) {
    if (init_value)
      set_all_valid();
    else
      clear();
  }
}

void Bitmask::set_all_valid(void) { memset(bitmask, 0x01, num_elements); }

void Bitmask::clear(void) { memset(bitmask, 0, num_elements); }

size_t Bitmask::count_set_bits(void) const
{
  size_t result = 0;
  for (size_t idx = 0; idx < num_elements; ++idx) result += get(idx);
  return result;
}

void Bitmask::copy(const Bitmask &target) const
{
  assert(num_elements == target.num_elements);
  memcpy(target.bitmask, bitmask, num_elements);
}

void intersect_bitmasks(Bitmask &out, const Bitmask &in1, const Bitmask &in2)
{
  assert(out.num_elements == in1.num_elements && out.num_elements == in2.num_elements);
  for (size_t idx = 0; idx < out.num_elements; ++idx) out.set(idx, in1.get(idx) && in2.get(idx));
}

}  // namespace pandas
}  // namespace legate
